src/core/pyproxy.c

```diff
@@ -1220,14 +1230,136 @@ EM_JS_NUM(int, pyproxy_init_js, (), {
   Module.wrapNamespace = function wrapNamespace(ns){
     return new Proxy(ns, NamespaceProxyHandlers);
   };
-
   return 0;
 });
 // clang-format on
 
+EM_JS_REF(JsRef, create_once_callback, (PyObject * obj), {
+  _Py_IncRef(obj);
+  let alreadyCalled = false;
+  function wrapper(... args)
+  {
+    if (alreadyCalled) {
+      throw new Error("OnceProxy can only be called once");
+    }
+    alreadyCalled = true;
+    try {
+      return Module.callPyObject(obj, ... args);
+    } finally {
+      _Py_DecRef(obj);
+    }
+  }
+  wrapper.destroy = function()
+  {
+    if (alreadyCalled) {
+      throw new Error("OnceProxy has already been destroyed");
+    }
+    alreadyCalled = true;
+    _Py_DecRef(obj);
+  };
+  return Module.hiwire.new_value(wrapper);
+});
+
+static PyObject*
+create_once_callback_py(PyObject* _mod, PyObject* obj)
+{
+  JsRef ref = create_once_callback(obj);
+  PyObject* result = JsProxy_create(ref);
+  hiwire_decref(ref);
+  return result;
+}
+
+// clang-format off
+EM_JS_REF(JsRef, create_promise_handles, (
+  PyObject* handle_result, PyObject* handle_exception
+), {
+  if (handle_result) {
+    _Py_IncRef(handle_result);
+  }
+  if (handle_exception) {
+    _Py_IncRef(handle_exception);
+  }
+  let used = false;
+  function checkUsed(){
+    if (used) {
+      throw new Error("One of the promise handles has already been called.");
+    }
+  }
+  function destroy(){
+    checkUsed();
+    used = true;
+    if(handle_result){
+      _Py_DecRef(handle_result);
+    }
+    if(handle_exception){
+      _Py_DecRef(handle_exception)
+    }
+  }
+  function onFulfilled(res) {
+    checkUsed();
+    try {
+      if(handle_result){
+        return Module.callPyObject(handle_result, res);
+      }
+    } finally {
+      destroy();
+    }
+  }
+  function onRejected(err) {
+    checkUsed();
+    try {
+      if(handle_exception){
+        return Module.callPyObject(handle_exception, err);
+      }
+    } finally {
+      destroy();
+    }
+  }
+  onFulfilled.destroy = destroy;
+  onRejected.destroy = destroy;
+  return Module.hiwire.new_value(
+    [onFulfilled, onRejected]
+  );
+})
+// clang-format on
+
+static PyObject*
+create_proxy(PyObject* _mod, PyObject* obj)
+{
+  JsRef ref = pyproxy_new(obj);
+  PyObject* result = JsProxy_create(ref);
+  hiwire_decref(ref);
+  return result;
+}
+
+static PyMethodDef pyproxy_methods[] = {
+  {
+    "create_once_callback",
+    create_once_callback_py,
+    METH_O,
+    PyDoc_STR(
+      "create_once_callback(obj : Callable) -> JsProxy"
```

Comment:
Maybe,
```suggestion
      "create_proxy_once(obj : Callable) -> JsProxy"
```
 to be more consistent with the one below? For me a callback is something that is meant to be called repeatedly, and even if we were in pure Python it's strange to have a function that creates a callback object. There is nothing that distinguishes a callback function from a non callback function generally (or in this case a Promise like object), unless I'm missing something? 

Also the docstring should mention that then and finally methods?

Reply:
I guess I meant that the argument should be a function of some sort, and the resulting object can only be used to call the function as opposed to `create_proxy` which makes sense for any Python argument and creates a proxy that can be used in other ways. I'm not sure what the best way to indicate this in the names though. `create_once_proxy` sounds better to me for word order than `create_proxy_once` though.

What about `create_once_callable`?

Reply:
> What about create_once_callable?

Sounds good.